docs/tracing.md

```diff
@@ -0,0 +1,237 @@
+for arguments of autodiff transformations, carry `ConcreteArray` abstract values
+that actually include the regular array data, and are used, e.g., for resolving
+conditionals. We will refer here to such tracers as **concrete tracers**. Tracer
+values computed from these concrete tracers, perhaps in combination with regular
+values, result in concrete tracers. A **concrete value** is either a regular
+value or a concrete tracer.
+
+Typically, computations that involve at least a tracer value will produce a
+tracer value. There are very few exceptions, when a computation can be
+entirely done using the abstract value carried by a tracer, in which case the
+result can be a **regular** Python value. For example, getting the shape of a
+tracer with `ShapedArray` abstract value. Another example is when explicitly
+casting a concrete tracer value to a regular type, e.g., `int(x)` or
+`x.astype(float)`. Another such situation is for `bool(x)`, which produces a
+Python bool when concreteness makes it possible. That case is especially salient
+because of how often it arises in control flow.
+
+Here is how the transformations introduce abstract or concrete tracers:
+
+* {func}`jax.jit`: introduces **abstract tracers** for all positional arguments
+  except those denoted by `static_argnums`, which remain regular
+  values.
+* {func}`jax.pmap`: introduces **abstract tracers** for all positional arguments
+  except those denoted by `static_broadcasted_argnums`.
+* {func}`jax.vmap`, {func}`jax.make_jaxpr`, {func}`xla_computation`:
+  introduce **abstract tracers** for all positional arguments.
+* {func}`jax.jvp` and {func}`jax.grad` introduce **concrete tracers**
+  for all positional arguments. An exception is when these transformations
+  are within an outer transformation and the actual arguments are
+  themselves abstract tracers; in that case, the tracers introduced
+  by the autodiff transformations are also abstract tracers.
+* All higher-order control-flow primitives ({func}`lax.cond`,
+  {func}`lax.while_loop`, {func}`lax.fori_loop`, {func}`lax.scan`) when they
+  process the functionals introduce **abstract tracers**, whether or not there
+  is a JAX transformation in progress.
+
+All of this is relevant when you have code that can operate
+only on regular Python values, such as code that has conditional
+control-flow based on data:
+
+```{code-cell}
+def divide(x, y):
+  return x / y if y >= 1. else 0.
+```
+
+If we want to apply {func}`jax.jit`, we must ensure to specify `static_argnums=1`
+to ensure `y` stays a regular value. This is due to the boolean expression
+`y >= 1.`, which requires concrete values (regular or tracers). The
+same would happen if we write explicitly `bool(y >= 1.)`, or `int(y)`,
+or `float(y)`.
+
+Interestingly, `jax.grad(divide)(3., 2.)`, works because {func}`jax.grad`
+uses concrete tracers, and resolves the conditional using the concrete
+value of `y`.
+
+### Static vs traced operations
```

Comment:
If it sounds good to you, let's swap the order of this and "Different kinds of JAX values" above. I think this section would flow a little better after the content on static args, and "Different kinds of JAX values" feels more esoteric/less crucial than static vs. traced operations (and now that I look closer, the text of Different Kinds is out of date -- `ConcreteArray` doesn't exist anymore. Let's leave that alone in this PR though, I don't totally understand the nuances of the updated tracing machinery.)